spec/eof.md

```diff
@@ -161,11 +159,28 @@ Code executing within an EOF environment will behave differently than legacy cod
 - The instruction `JUMPDEST` is renamed to `NOP` and remains charging 1 gas without any effect.
     - Note: jumpdest-analysis is not performed anymore.
 - EOF contract may not deploy legacy code (it is naturally rejected on the code validation stage)
-- Legacy creation transactions (any tranactions with empty `to`) are invalid in case `data` contains EOF code (starts with `EF00` magic)
 - If instructions `CREATE` and `CREATE2` have EOF code as initcode (starting with `EF00` magic)
     - deployment fails (returns 0 on the stack)
     - caller's nonce is not updated and gas for initcode execution is not consumed
 
+#### Creation transactions
+
+Creation transactions (tranactions with empty `to`), with `data` containing EOF code (starting with `EF00` magic) are interpreted as having an EOF `initcontainer` in the `data` and:
+
+- intrinsic gas cost rules and limits defined in EIP-3860 for legacy creation transaction apply. The entire `data` of the transaction is used for these calculations
+- **the initcontainer and all its subcontainers are validated recursively**
+- transaction fails if any of those checks were invalid (gas for initcode execution is not consumed. Only intrinsic creation transaction costs were consumed)
+- bytes after those declared as part of the initcode container's sections', are treated as calldata to pass into the execution frame
```

Comment:
Why after container sections? I think there also may be data section, and there should be extra requirement that it's declared size equals actual size.

Reply:
Sorry, unfortunate wording, I meant all sections of the container :sweat_smile: . Will fix that

Reply:
I pushed a (hopefully) less ambiguous wording

> there should be extra requirement that it's declared size equals actual size.

Wait - I think it goes the other way around - we need to assume it has the declared size, as everything after that is assumed to be the calldata appended?

Reply:
Yes it's the same as what I meant, my idea was just our general validation algorithm allows that `the total size of not yet deployed container might be up to data_size lower than the above values...`, so I think it should be mentioned somehow at validation step, that declared data size == actual size.